lua-loader/lua-ckb.c

```
@@ -0,0 +1,281 @@
+  size_t size;
+  int integer;
+} FIELD_ARG;
+
+typedef struct {
+  string name;
+  FIELD_TYPE type;
+  FIELD_ARG arg;
+} FIELD;
+
+/////////////////////////////////////////////////////
+// Utilities
+/////////////////////////////////////////////////////
+
+#define THROW_ERROR(L, s, ...)                                                 \
+  char _error[256];                                                            \
+  snprintf_(_error, 256, s, __VA_ARGS__);                                      \
+  lua_pushstring(L, _error);                                                   \
+  lua_error(L);
+
+#define CALL_SYSCALL_PUSH_RESULT(L, f, l, ...)                                 \
+  int _ret = 0;                                                                \
+  uint8_t *_buf = NULL;                                                        \
+  if (l == 0) {                                                                \
+    /* just get buffer length */                                               \
+    _ret = f(NULL, &l, __VA_ARGS__);                                           \
+  } else {                                                                     \
+    _buf = malloc(l);                                                          \
+  }                                                                            \
+  if (_ret == 0) {                                                             \
+    if (_buf == NULL) {                                                        \
+      /* malloc for an empty buffer */                                         \
+      _buf = malloc(l);                                                        \
+    }                                                                          \
+    _ret = f(_buf, &l, __VA_ARGS__);                                           \
+  }                                                                            \
+  if (_ret != 0) {                                                             \
```

It should return the raw error code to Lua world.